File: src/main/java/fi/quanfoxes/parser/nodes/LoopNode.java
package fi.quanfoxes.parser.nodes;

import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Node;

public class LoopNode extends Node {
    private Context context;

    public LoopNode(Context context, Node condition, Node body) {
        this.context = context;

        if (condition != null) {
            super.add(condition);
        }

        super.add(body);
    }

    /**
     * Returns the private context of the while statement
     * @return Private context of the while statement
     */
    public Context getContext() {
        return context;
    }

    /**
     * Returns whether this loop loops forever
     * @return True if this loop loops forever, otherwise false
     */
    public boolean isForever() {
        return first() == last();
    }

    /**
     * Returns the node that is executed when program enters the while statement
     * @return Node that is executed when program enters the while statement
     */
    public Node getStart() {
        return first().first();
    }

    /**
     * Returns the node that represents the condition of the statement
     * @return
     */
    public Node getCondition() {
        return getStart().next();
    }

    /**
     * Returns the node that represents the action that is done in the end of each cycle in the while statement
     * @return Node that represents the action that is done in the end of each cycle in the while statement
     */
    public Node getAction() {
        return first().last();
    }

    /**
     * Returns the node that is executed each cycle in the while statement
     * @return Node that is executed each cycle in the while statement
     */
    public Node getBody() {
        return last();
    }
}

File: Vivid/Parser/Nodes/LoopNode.cs
using System;

public class LoopNode : Node, IResolvable, IScope
{
	public Context Context { get; private set; }

	public Node Steps => First!;
	public ScopeNode Body => Last!.To<ScopeNode>();

	public Node Initialization => Steps.First!;
	public Node Condition => Common.FindCondition(Initialization.Next!);
	public Node Action => Steps.Last!;

	public Label? Start { get; set; } = null;
	public Label? Exit { get; set; } = null;

	public bool IsForeverLoop => First == Last;

	public LoopNode(Context context, Node? steps, ScopeNode body, Position? position = null)
	{
		Context = context;
		Position = position;
		Instance = NodeType.LOOP;

		if (steps != null)
		{
			Add(steps);
		}

		Add(body);
	}

	/// <summary>
	/// Returns the nodes which are executed during condition step except the actual condition.
	/// NOTE: Use this function only for building since this function returns copies of the executed nodes
	/// </summary>
	public Node GetConditionInitialization()
	{
		// Clone all the nodes under the condition step
		var node = Initialization.Next!.Clone();
		node.Parent = this;

		// Remove the condition from the initialization since it is built separately
		if (!Common.FindCondition(node).Remove())
		{
			throw new ApplicationException("Could not remove the condition from the condition step initialization");
		}

		return node;
	}

	/// <summary>
	/// Returns all the nodes which are executed during the condition step
	/// </summary>
	public Node GetConditionContainer()
	{
		return Initialization.Next!;
	}

	public Node? Resolve(Context context)
	{
		if (!IsForeverLoop)
		{
			Resolver.Resolve(Context, Initialization);
			Resolver.Resolve(Context, Condition);
			Resolver.Resolve(Context, Action);
		}

		Resolver.Resolve(Body.Context, Body);

		return null;
	}

	public Status GetStatus()
	{
		return Status.OK;
	}

	public void SetContext(Context context)
	{
		Context = context;
	}

	public Context GetContext()
	{
		return Context;
	}

	public override int GetHashCode()
	{
		return HashCode.Combine(Instance, Position, Context.Identity, Body.Context.Identity, IsForeverLoop);
	}

	public override string ToString() => "Loop";
}
